File: ios/Classes/models/map/here_map/MapController.swift
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import CarPlay
import heresdk
import UIKit

enum ConstantsEnum {
    static let DEFAULT_MAP_CENTER = GeoCoordinates(latitude: 52.520798, longitude: 13.409408)
    static let DEFAULT_DISTANCE_IN_METERS: Double = 1000 * 2
}

// An app that allows to calculate a route and start navigation, using either platform positioning or
// simulated locations.
class MapController: LongPressDelegate {
    private let viewController: UIViewController
    private let mapView: MapView
    private let routeCalculator: RouteCalculator
    private let navigationHelper: NavigationHelper
    private var mapMarkers = [MapMarker]()
    private var mapPolygons = [MapPolygon]()
    private var mapPolylineList = [MapPolyline]()
    private var startingWaypoint: Waypoint?
    private var destinationWaypoint: Waypoint?
    private var isLongpressDestination = false
    private var messageTextView: UITextView

    /// Initialize the MapController
    /// - Parameters:
    ///   - viewController: ViewController that holds the MapController
    ///   - mapView: MapView that renders the map
    ///   - messageTextView: TextView that displays log messages
    init(viewController: UIViewController, mapView: MapView, messageTextView: UITextView) {
        self.viewController = viewController
        self.mapView = mapView
        self.messageTextView = messageTextView

        let distanceInMeters = MapMeasure(kind: .distance, value: ConstantsEnum.DEFAULT_DISTANCE_IN_METERS)
        mapView.camera.lookAt(point: ConstantsEnum.DEFAULT_MAP_CENTER,
                              zoom: distanceInMeters)

        routeCalculator = RouteCalculator()

        navigationHelper = NavigationHelper(mapView: mapView,
                                            messageTextView: messageTextView)
        navigationHelper.startLocationProvider()

        setLongPressGestureHandler()
        showMessage("Long press to set a destination or use a random one.")
    }

    /// Set destination waypoint for route calculation.
    func setDestinationWaypoint(_ destination: Waypoint) {
        destinationWaypoint = destination
    }

    /// Set the normalized principal point of the VisualNavigator camera.
    func setVisualNavigatorCameraPoint(normalizedPrincipalPoint: heresdk.Anchor2D) {
        navigationHelper.setVisualNavigatorCameraPoint(normalizedPrincipalPoint: normalizedPrincipalPoint)
    }

    /// Calculate a route and start navigation using a location simulator.
    /// Start is map center and destination location is set random within viewport,
    /// unless a destination is set via long press.
    func addRouteSimulatedLocation() {
        calculateRoute(isSimulated: true)
    }

    /// Calculate a route and start navigation using locations from device.
    /// Start is current location and destination is set random within viewport,
    /// unless a destination is set via long press.
    func addRouteDeviceLocation() {
        calculateRoute(isSimulated: false)
    }

    /// Clears the map and resets the route and navigation.
    func clearMapButtonClicked() {
        clearMap()
        isLongpressDestination = false
    }

    /// Enable or disable camera tracking
    func enableCameraTracking() {
        navigationHelper.startCameraTracking()
    }

    /// Enable or disable camera tracking
    func disableCameraTracking() {
        navigationHelper.stopCameraTracking()
    }

    func getIsNavigationInProgress() -> Bool {
        return navigationHelper.getIsNavigationInProgress()
    }

    /// Add a marker on the map.
    /// - Parameters:
    ///   - coordinates: The coordinates of the marker
    ///   - markerImage: The image of the marker
    func addMapMarker(coordinates: GeoCoordinates, markerImage: UIImage, markerSize: CGSize, metadata: heresdk.Metadata?) {
        let marker = mapMarkers.first(where: {
            $0.metadata?.getString(key: "marker") == metadata?.getString(key: "marker")
        })

        if let existingMarker = marker {
            existingMarker.coordinates = coordinates
        } else {
            guard let imageData = markerImage.pngData() else { return }

            let mapImage = MapImage(imageData: imageData,
                                    imageFormat: ImageFormat.png, width: UInt32(markerSize.width), height: UInt32(markerSize.height))
            let mapMarker = MapMarker(at: coordinates,
                                      image: mapImage)
            mapMarker.metadata = metadata
            mapView.mapScene.addMapMarker(mapMarker)
            mapMarkers.append(mapMarker)
        }
    }

    /// Add a polygon on the map.
    /// - Parameters:
    ///   - coordinate: The coordinates of the polygon
    ///   - accuracy: The accuracy of the polygon
    ///   - metadata: The metadata of the polygon
    func addMapPolygon(coordinate: GeoCoordinates, accuracy: Double, metadata: heresdk.Metadata?) {
        let polygon = mapPolygons.first(where: {
            $0.metadata?.getString(key: "polygon") == metadata?.getString(key: "polygon")
        })

        if let existingpolygon = polygon {
            existingpolygon.geometry = GeoPolygon(geoCircle: GeoCircle(center: coordinate, radiusInMeters: accuracy))
        } else {
            let mapPolygon = MapPolygon(geometry: GeoPolygon(geoCircle: GeoCircle(center: coordinate, radiusInMeters: accuracy)), color: UIColor(argb: 0x550B_C7C2))
            mapPolygon.metadata = metadata
            mapView.mapScene.addMapPolygon(mapPolygon)
            mapPolygons.append(mapPolygon)
        }
    }

    /// Calculates a route and start navigation.
    /// - Parameter isSimulated: Whether to use simulated locations.
    private func calculateRoute(isSimulated: Bool) {
        clearMap()

        if !determineRouteWaypoints(isSimulated: isSimulated) {
            return
        }

        // Calculates a car route.
        routeCalculator.calculateRoute(start: startingWaypoint!,
                                       destination: destinationWaypoint!)
        { routingError, routes in
            if let error = routingError {
                self.onNavigationFailed(title: "Error while calculating a route:", message: "\(error)")
                return
            }

            // When routingError is nil, routes is guaranteed to contain at least one route.
            let route = routes!.first
            self.showRouteOnMap(route: route!)
            self.showRouteDetails(route: route!, isSimulated: isSimulated)
        }
    }

    /// Determines the starting and destination waypoints for the route calculation.
    /// - Parameter isSimulated: Whether to use simulated locations.
    /// - Returns: True if waypoints are determined, false if not.
    private func determineRouteWaypoints(isSimulated: Bool) -> Bool {
        // When using real GPS locations, we always start from the current location of user.
        guard let location = navigationHelper.getLastKnownLocation() else {
            onNavigationFailed(title: "Error", message: "No location found.")
            return false
        }
        print("Last known location: \(location.coordinates)")

        startingWaypoint = Waypoint(coordinates: location.coordinates)

        // If a driver is moving, the bearing value can help to improve the route calculation.
        startingWaypoint?.headingInDegrees = location.bearingInDegrees

        mapView.camera.lookAt(point: location.coordinates)

        if isSimulated {
            if destinationWaypoint == nil {
                destinationWaypoint = Waypoint(coordinates: createRandomGeoCoordinatesAroundMapCenter())
            }
        }
        return true
    }

    /// Shows the route details on the screen.
    /// - Parameters:
    ///   - route: The route.
    ///   - isSimulated: Whether to use simulated locations.
    private func showRouteDetails(route: Route, isSimulated: Bool) {
        let estimatedTravelTimeInSeconds = route.duration
        let lengthInMeters = route.lengthInMeters

//        let routeDetails = "Travel Time: " + formatTime(sec: estimatedTravelTimeInSeconds)
//                         + ", Length: " + formatLength(meters: lengthInMeters)

        navigationHelper.startNavigation(route: route, isSimulated: isSimulated)

        // showStartNavigationDialog(title: "Route Details",
//                                  message: routeDetails,
//                                  route: route,
//                                  isSimulated: isSimulated)
    }

    /// Format time in minutes and seconds
    /// - Parameter sec: Time in seconds
    /// - Returns: Time in minutes and seconds
    private func formatTime(sec: Double) -> String {
        let hours: Double = sec / 3600
        let minutes: Double = sec.truncatingRemainder(dividingBy: 3600) / 60

        return "\(Int32(hours)):\(Int32(minutes))"
    }

    /// Format length in kilometers and meters
    /// - Parameter meters: Length in meters
    /// - Returns: Length in kilometers and meters
    private func formatLength(meters: Int32) -> String {
        let kilometers: Int32 = meters / 1000
        let remainingMeters: Int32 = meters % 1000

        return "\(kilometers).\(remainingMeters) km"
    }

    /// Show route on the map
    /// - Parameter route: Route
    private func showRouteOnMap(route: Route) {
        // Show route as polyline.
        let routeGeoPolyline = route.geometry
        let widthInPixels = 20.0
        let polylineColor = UIColor(red: 0, green: 0.56, blue: 0.54, alpha: 0.63)
        do {
            let routeMapPolyline = try MapPolyline(geometry: routeGeoPolyline,
                                                   representation: MapPolyline.SolidRepresentation(
                                                       lineWidth: MapMeasureDependentRenderSize(
                                                           sizeUnit: RenderSize.Unit.pixels,
                                                           size: widthInPixels
                                                       ),
                                                       color: polylineColor,
                                                       capShape: LineCap.round
                                                   ))
            mapView.mapScene.addMapPolyline(routeMapPolyline)
            mapPolylineList.append(routeMapPolyline)
        } catch {
            fatalError("Failed to render MapPolyline. Cause: \(error)")
        }
    }

    /// Clear the map
    func clearMap() {
        clearWaypointMapMarkers()
        clearMapPolygons()
        clearRoute()

        navigationHelper.stopNavigation()
    }

    /// Clear the map markers
    private func clearWaypointMapMarkers() {
        for mapMarker in mapMarkers {
            mapView.mapScene.removeMapMarker(mapMarker)
        }
        mapMarkers.removeAll()
    }

    /// Clear the map polygons
    private func clearMapPolygons() {
        for mapPolygon in mapPolygons {
            mapView.mapScene.removeMapPolygon(mapPolygon)
        }
        mapPolygons.removeAll()
    }

    /// Clear the route
    private func clearRoute() {
        for mapPolyline in mapPolylineList {
            mapView.mapScene.removeMapPolyline(mapPolyline)
        }
        mapPolylineList.removeAll()
    }

    /// Set the long press gesture handler
    private func setLongPressGestureHandler() {
        mapView.gestures.longPressDelegate = self
    }

    /// Conform to LongPressDelegate protocol .
    /// - Parameters:
    ///   - state: The state of the gesture.
    ///   - origin: The point where the gesture began.
    func onLongPress(state: GestureState, origin: Point2D) {
        guard let geoCoordinates = mapView.viewToGeoCoordinates(viewCoordinates: origin) else {
            print("Warning: Long press coordinate is not on map view.")
            return
        }

        if state == GestureState.begin {
            if isLongpressDestination {
                destinationWaypoint = Waypoint(coordinates: geoCoordinates)
                addCircleMapMarker(geoCoordinates: destinationWaypoint!.coordinates, imageName: "green_dot.png")
                showMessage("New long press destination set.")
            } else {
                startingWaypoint = Waypoint(coordinates: geoCoordinates)
                addCircleMapMarker(geoCoordinates: startingWaypoint!.coordinates, imageName: "green_dot.png")
                showMessage("New long press starting point set.")
            }
            isLongpressDestination = !isLongpressDestination
        }
    }

    /// Create a random geo coordinates
    /// - Returns: Random geo coordinates
    private func createRandomGeoCoordinatesAroundMapCenter() -> GeoCoordinates {
        let centerGeoCoordinates = getMapViewCenter()
        let lat = centerGeoCoordinates.latitude
        let lon = centerGeoCoordinates.longitude
        return GeoCoordinates(latitude: getRandom(min: lat - 0.02,
                                                  max: lat + 0.02),
                              longitude: getRandom(min: lon - 0.02,
                                                   max: lon + 0.02))
    }

    /// Get random number
    /// - Parameters:
    ///   - min: Minimum number
    ///   - max: Maximum number
    /// - Returns: Random number
    private func getRandom(min: Double, max: Double) -> Double {
        return Double.random(in: min ... max)
    }

    /// Get the map view center
    /// - Returns: Map view center
    private func getMapViewCenter() -> GeoCoordinates {
        return mapView.camera.state.targetCoordinates
    }

    /// Get the last known location
    /// - Returns: Last known location
    func getLastKnownLocation() -> Location? {
        return navigationHelper.getLastKnownLocation()
    }

    /// Get the marker coordinates
    /// - Parameter markerType: Map Marker type
    /// - Returns: Marker coordinates
    func getMarkerCoordinates(markerType: MapMarkerType) -> GeoCoordinates? {
        let marker = mapMarkers.first(where: {
            $0.metadata?.getString(key: "marker") == markerType.rawValue
        })

        if let existingMarker = marker {
            return existingMarker.coordinates
        }

        return nil
    }

    /// Remove a marker
    /// - Parameter markerType: Map Marker type
    func removeMarker(markerType: MapMarkerType) {
        let marker = mapMarkers.first(where: {
            $0.metadata?.getString(key: "marker") == markerType.rawValue
        })

        if let existingMarker = marker {
            mapView.mapScene.removeMapMarker(existingMarker)
            mapMarkers.removeAll(where: { $0.metadata?.getString(key: "marker") == markerType.rawValue })
        }
    }

    /// Remove a polygon
    /// - Parameter markerType: Map Marker type
    func removePolygon(markerType: MapMarkerType) {
        let polygon = mapPolygons.first(where: {
            $0.metadata?.getString(key: "polygon") == markerType.rawValue
        })

        if let existingPolygon = polygon {
            mapView.mapScene.removeMapPolygon(existingPolygon)
            mapPolygons.removeAll(where: { $0.metadata?.getString(key: "polygon") == markerType.rawValue })
        }
    }

    /// Add a circle map marker
    /// - Parameters:
    ///   - geoCoordinates: Geo coordinates
    ///   - imageName: Image name
    private func addCircleMapMarker(geoCoordinates: GeoCoordinates, imageName: String) {
        guard
            let image = UIImage(named: imageName),
            let imageData = image.pngData()
        else {
            return
        }

        let mapImage = MapImage(pixelData: imageData,
                                imageFormat: ImageFormat.png)
        let mapMarker = MapMarker(at: geoCoordinates,
                                  image: mapImage)
        mapView.mapScene.addMapMarker(mapMarker)
        mapMarkers.append(mapMarker)
    }

    /// Show dialog with navigation options
    /// - Parameters:
    ///   - title: Title of the dialog
    ///   - message: Message of the dialog
    ///   - route: Route to be started
    ///   - isSimulated: Whether to use simulated locations
    private func showStartNavigationDialog(title: String,
                                           message: String,
                                           route: Route,
                                           isSimulated: Bool)
    {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let buttonText = isSimulated ? "Start navigation (simulated)" : "Start navigation (device location)"
        alertController.addAction(UIAlertAction(title: buttonText, style: .default, handler: { _ in
            self.navigationHelper.startNavigation(route: route, isSimulated: isSimulated)
        }))
        alertController.addAction(UIAlertAction(title: "Cancel", style: .default))
        viewController.present(alertController, animated: true)
    }

    /// Show dialog with title and message
    /// - Parameters:
    ///   - title: Title of the dialog
    ///   - message: Message of the dialog
    private func onNavigationFailed(title: String, message: String) {
        debugPrint("\(title) => \(message)")

        FCPStreamHandlerPlugin.sendEvent(type: FCPChannelTypes.onNavigationFailedFromCarPlay,
                                         data: [
                                             "message": message,
                                         ])
//        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
//        alertController.addAction(UIAlertAction(title: "OK", style: .default))
//        viewController.present(alertController, animated: true)
    }

    /// A permanent view to show log content.
    /// - Parameter message: Log message to show in the view.
    private func showMessage(_ message: String) {
        messageTextView.text = message
        messageTextView.textColor = .white
        messageTextView.layer.cornerRadius = 8
        messageTextView.isEditable = false
        messageTextView.textAlignment = NSTextAlignment.center
        messageTextView.font = .systemFont(ofSize: 14)
    }
}

File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/map/here_map/MapController.kt
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.oguzhnatly.flutter_carplay.models.map.here_map

import android.annotation.SuppressLint
import com.here.sdk.core.Color
import com.here.sdk.core.GeoCircle
import com.here.sdk.core.GeoCoordinates
import com.here.sdk.core.GeoPolygon
import com.here.sdk.core.GeoPolyline
import com.here.sdk.core.Location
import com.here.sdk.core.Metadata
import com.here.sdk.core.Point2D
import com.here.sdk.core.engine.SDKNativeEngine
import com.here.sdk.gestures.GestureState
import com.here.sdk.gestures.LongPressListener
import com.here.sdk.mapview.ImageFormat
import com.here.sdk.mapview.LineCap
import com.here.sdk.mapview.MapImage
import com.here.sdk.mapview.MapMarker
import com.here.sdk.mapview.MapMeasure
import com.here.sdk.mapview.MapMeasureDependentRenderSize
import com.here.sdk.mapview.MapPolygon
import com.here.sdk.mapview.MapPolyline
import com.here.sdk.mapview.MapSurface
import com.here.sdk.mapview.RenderSize
import com.here.sdk.routing.Route
import com.here.sdk.routing.Waypoint
import com.oguzhnatly.flutter_carplay.Bool
import com.oguzhnatly.flutter_carplay.FCPChannelTypes
import com.oguzhnatly.flutter_carplay.FCPStreamHandlerPlugin
import com.oguzhnatly.flutter_carplay.Logger
import com.oguzhnatly.flutter_carplay.MapMarkerType
import com.oguzhnatly.flutter_carplay.UIImage
import java.util.Locale

object ConstantsEnum {
    val DEFAULT_MAP_CENTER = GeoCoordinates(52.520798, 13.409408)
    val DEFAULT_DISTANCE_IN_METERS = 1000.0 * 2
    val ROUTE_DEVIATION_DISTANCE = 20.0 // In Meters
}

data class CGSize(val width: Double, val height: Double)

/**
 * An app that allows to calculate a route and start navigation, using either platform positioning
 * or simulated locations.
 */
class MapController(private val mapView: MapSurface) {
    private val mapMarkers: MutableList<MapMarker> = ArrayList()
    private val mapPolygons: MutableList<MapPolygon> = ArrayList()
    private val mapPolylineList: MutableList<MapPolyline?> = ArrayList()
    private var startWaypoint: Waypoint? = null
    private var destinationWaypoint: Waypoint? = null
    private var setLongpressDestination = false
    private var routeCalculator: RouteCalculator
//    private var isCameraTrackingEnabled = true

    /// NavigationHelper instance
    val navigationHelper: NavigationHelper

    /// lastKnownLocation getter
    val lastKnownLocation: Location?
        get() = navigationHelper.lastKnownLocation

    init {
        val distanceInMeters =
            MapMeasure(MapMeasure.Kind.DISTANCE, ConstantsEnum.DEFAULT_DISTANCE_IN_METERS)
        this.mapView.camera.lookAt(ConstantsEnum.DEFAULT_MAP_CENTER, distanceInMeters)

        routeCalculator =
            RouteCalculator(
                SDKNativeEngine.getSharedInstance()?.isOfflineMode ?: false
            )

        navigationHelper = NavigationHelper(mapView)
        navigationHelper.startLocationProvider()

        // Toggle offline mode to change the RouteCalculator instance
        toggleOfflineModeHandler = { isOffline: Bool ->
            routeCalculator = RouteCalculator(isOffline)
        }

        // Re-Routing callback to find new route
        reroutingHandler = { startWayPoint, completion ->

//            val navigationSession =
//                (FlutterCarplayPlugin.fcpRootTemplate as? FCPMapTemplate)?.navigationSession

//            if #available(iOS 15.4, *) {
//                navigationSession?.pauseTrip(for: .rerouting, description: "", turnCardColor: .systemGreen)
//            } else {
//                navigationSession?.pauseTrip(for: .rerouting, description: "")
//            }

            startWaypoint = startWayPoint

            // Calculates a car route.
            routeCalculator.calculateRoute(
                start = startWayPoint,
                destination = this.destinationWaypoint!!
            )
            { routingError, routes ->
                routingError?.let {
                    onNavigationFailed(
                        title = "Error while calculating a reroute =",
                        message = "$it"
                    )
                    completion()
                    return@calculateRoute
                }

                // When routingError is nil, routes is guaranteed to contain at least one route.
                val route = routes!!.first()
                showRouteDetails(route!!)
                navigationHelper.onNewRoute(route)
                completion()
            }
        }

        setLongPressGestureHandler()

    }

    /** Set destination waypoint for route calculation. */
    fun setDestinationWaypoint(destination: Waypoint) {
        destinationWaypoint = destination
    }

    /**
     * Calculate a route and start navigation using a location simulator.
     * Start is map center and destination location is set random within viewport,
     * unless a destination is set via long press.
     */
    fun addRouteSimulatedLocation() {
        calculateRoute(true)
    }

    /**
     * Calculate a route and start navigation using locations from the device.
     * Start is the current location and the destination is set randomly within the viewport,
     * unless a destination is set via long press.
     */
    fun addRouteDeviceLocation() {
        calculateRoute(false)
    }

    /**
     * Add a marker on the map.
     *
     * @param coordinates The coordinates of the marker
     * @param markerImage The image of the marker
     * @param markerSize The size of the marker
     * @param metadata The metadata of the marker
     */
    @SuppressLint("RestrictedApi")
    fun addMapMarker(
        coordinates: GeoCoordinates,
        markerImage: UIImage,
        markerSize: CGSize,
        metadata: Metadata?,
    ) {
        val marker = mapMarkers.find {
            it.metadata?.getString("marker") == metadata?.getString("marker")
        }

        if (marker != null) {
            marker.coordinates = coordinates
            mapView.mapScene.removeMapMarker(marker)
            mapView.mapScene.addMapMarker(marker)
        } else {
            val imageData = markerImage.icon?.mData ?: return
            val mapImage = MapImage(
                imageData,
                ImageFormat.PNG,
                markerSize.width.toLong(),
                markerSize.height.toLong(),
            )
            val mapMarker = MapMarker(coordinates, mapImage)
            mapMarker.metadata = metadata
            mapView.mapScene.addMapMarker(mapMarker)
            mapMarkers.add(mapMarker)
        }
    }

    /**
     * Add a polygon on the map.
     *
     * @param coordinate The coordinates of the polygon
     * @param accuracy The accuracy of the polygon
     * @param metadata The metadata of the polygon
     */
    fun addMapPolygon(coordinate: GeoCoordinates, accuracy: Double, metadata: Metadata?) {
        val polygon = mapPolygons.find {
            it.metadata?.getString("polygon") == metadata?.getString("polygon")
        }

        if (polygon != null) {
            polygon.geometry = GeoPolygon(GeoCircle(coordinate, accuracy))
            mapView.mapScene.removeMapPolygon(polygon)
            mapView.mapScene.addMapPolygon(polygon)
        } else {
            val mapPolygon = MapPolygon(
                GeoPolygon(GeoCircle(coordinate, accuracy)),
                Color.valueOf(0x550BC7C2)
            )
            mapPolygon.metadata = metadata
            mapView.mapScene.addMapPolygon(mapPolygon)
            mapPolygons.add(mapPolygon)
        }
    }

    /**
     * Calculates a route and starts navigation.
     *
     * @param isSimulated Whether to use simulated locations.
     */
    private fun calculateRoute(isSimulated: Boolean) {
        clearMap(clearInitialOnly = true)

        if (!determineRouteWaypoints(isSimulated)) return

        // Calculates a car route.
        routeCalculator.calculateRoute(startWaypoint, destinationWaypoint) { routingError, routes ->
            if (routingError == null) {
                // When routingError is null, routes is guaranteed to contain at least one route.
                showRouteDetails(routes!!.first())
            } else {
                onNavigationFailed(
                    title = "Error while calculating a route:",
                    message = "$routingError"
                )
            }
        }
    }

    /**
     * Determines the starting and destination waypoints for the route calculation.
     *
     * @param isSimulated Whether to use simulated locations.
     * @return True if waypoints are determined, false if not.
     */
    private fun determineRouteWaypoints(isSimulated: Boolean): Boolean {
        // When using real GPS locations, we always start from the current location of user.
        val location = lastKnownLocation

        if (location == null) {
            onNavigationFailed(title = "Error", message = "No GPS location found.")
            return false
        }

        startWaypoint = Waypoint(location.coordinates)

        // If a driver is moving, the bearing value can help to improve the route calculation.
        startWaypoint?.headingInDegrees = location.bearingInDegrees

        // Update the camera position.
        mapView.camera.lookAt(location.coordinates)

        // When using real GPS locations, we always start from the current location of user.
        if (isSimulated && destinationWaypoint == null) {
            destinationWaypoint = Waypoint(createRandomGeoCoordinatesAroundMapCenter())
        }

        return true
    }

    /**
     * Shows the route details on the screen.
     *
     * @param route The route.
     */
    private fun showRouteDetails(route: Route) {
        val estimatedTravelTimeInSeconds: Long = route.duration.seconds
        val lengthInMeters: Int = route.lengthInMeters

        val routeDetails =
            ("Travel Time: " +
                    formatTime(estimatedTravelTimeInSeconds) +
                    ", Length: " +
                    formatLength(lengthInMeters))

//        showStartNavigationDialog(
//            "Route Details",
//            routeDetails,
//            route,
//            isSimulated
//        )
    }

    /**
     * Format time in minutes and seconds.
     *
     * @param sec the time in seconds
     * @return the time in minutes and seconds
     */
    private fun formatTime(sec: Long): String {
        val hours = (sec / 3600).toInt()
        val minutes = ((sec % 3600) / 60).toInt()

        return String.format(Locale.getDefault(), "%02d:%02d", hours, minutes)
    }

    /**
     * Format length in kilometers and meters.
     *
     * @param meters the length in meters
     * @return the length in kilometers and meters
     */
    private fun formatLength(meters: Int): String {
        val kilometers = meters / 1000
        val remainingMeters = meters % 1000

        return String.format(Locale.getDefault(), "%02d.%02d km", kilometers, remainingMeters)
    }

    /**
     * Shows the route on the map.
     *
     * @param route The route.
     */
    private fun showRouteOnMap(route: Route) {
        // Show route as polyline.
        val routeGeoPolyline: GeoPolyline = route.geometry
        val widthInPixels = 20f
        val polylineColor = Color.valueOf(0f, 0.56f, 0.54f, 0.63f)
        var routeMapPolyline: MapPolyline? = null
        try {
            routeMapPolyline =
                MapPolyline(
                    routeGeoPolyline,
                    MapPolyline.SolidRepresentation(
                        MapMeasureDependentRenderSize(
                            RenderSize.Unit.PIXELS,
                            widthInPixels.toDouble()
                        ),
                        polylineColor,
                        LineCap.ROUND
                    )
                )
        } catch (e: MapPolyline.Representation.InstantiationException) {
            Logger.log("MapPolyline Representation Exception:", e.error.name)
        } catch (e: MapMeasureDependentRenderSize.InstantiationException) {
            Logger.log("MapMeasureDependentRenderSize Exception:", e.error.name)
        }

        routeMapPolyline?.let { mapView.mapScene.addMapPolyline(it) }
        mapPolylineList.add(routeMapPolyline)
    }

    /**
     * Clears the map.
     *
     * @param clearInitialOnly Whether to clear only the initial map marker
     */
    fun clearMap(clearInitialOnly: Bool = false) {
        if (clearInitialOnly) {
            removeMarker(MapMarkerType.INITIAL)
            removePolygon(MapMarkerType.INITIAL)
        } else {
            clearWaypointMapMarkers()
            clearMapPolygons()
        }
        clearRoute()

        navigationHelper.stopNavigation()
    }

    /** Clear the map markers. */
    private fun clearWaypointMapMarkers() {
        for (mapMarker in mapMarkers) {
            mapView.mapScene.removeMapMarker(mapMarker)
        }
        mapMarkers.clear()
    }

    /** Clear the map polygons. */
    private fun clearMapPolygons() {
        for (mapPolygon in mapPolygons) {
            mapView.mapScene.removeMapPolygon(mapPolygon)
        }
        mapPolygons.clear()
    }

    /** Clear the route. */
    private fun clearRoute() {
        for (mapPolyline in mapPolylineList) {
            mapPolyline?.let { mapView.mapScene.removeMapPolyline(it) }
        }
        mapPolylineList.clear()
    }

    /** Set the long press gesture handler. */
    private fun setLongPressGestureHandler() {
        mapView.gestures.longPressListener =
            LongPressListener { gestureState: GestureState, touchPoint: Point2D? ->
                val geoCoordinates: GeoCoordinates =
                    touchPoint?.let { mapView.viewToGeoCoordinates(it) }
                        ?: return@LongPressListener
                if (gestureState == GestureState.BEGIN) {
                    if (setLongpressDestination) {
                        destinationWaypoint = Waypoint(geoCoordinates)
//                        addCircleMapMarker(geoCoordinates, R.drawable.green_dot)
//                        messageView.text = "New long press destination set."
                    } else {
                        startWaypoint = Waypoint(geoCoordinates)
//                        addCircleMapMarker(geoCoordinates, R.drawable.green_dot)
//                        messageView.text = "New long press starting point set."
                    }
                    setLongpressDestination = !setLongpressDestination
                }
            }
    }

    /** Create a random geo coordinates.
     *
     * @return the random geo coordinates
     */
    private fun createRandomGeoCoordinatesAroundMapCenter(): GeoCoordinates {
        val centerGeoCoordinates: GeoCoordinates = mapViewCenter
        val lat: Double = centerGeoCoordinates.latitude
        val lon: Double = centerGeoCoordinates.longitude
        return GeoCoordinates(getRandom(lat - 0.02, lat + 0.02), getRandom(lon - 0.02, lon + 0.02))
    }

    /**
     * Get a random number between the specified minimum and maximum values (inclusive).
     *
     * @param min The minimum number.
     * @param max The maximum number.
     * @return The randomly generated number.
     */
    private fun getRandom(min: Double, max: Double): Double {
        return min + Math.random() * (max - min)
    }

    /**
     * Get the map view center.
     *
     * @return The map view center.
     */
    private val mapViewCenter: GeoCoordinates
        get() = mapView.camera.state.targetCoordinates

    /**
     * Get the marker coordinates.
     *
     * @param markerType The type of the map marker.
     * @return The coordinates of the marker.
     */
    fun getMarkerCoordinates(markerType: MapMarkerType): GeoCoordinates? {
        val marker = mapMarkers.find {
            it.metadata?.getString("marker") == markerType.name
        }

        return marker?.coordinates
    }

    /**
     * Remove a marker.
     *
     * @param markerType The type of the map marker to be removed.
     */
    fun removeMarker(markerType: MapMarkerType) {
        val marker = mapMarkers.find {
            it.metadata?.getString("marker") == markerType.name
        }

        marker?.let { mapMarker ->
            mapView.mapScene.removeMapMarker(mapMarker)
            mapMarkers.removeAll { it.metadata?.getString("marker") == markerType.name }
        }
    }

    /**
     * Removes a polygon from the map.
     *
     * @param markerType The type of the map marker.
     */
    fun removePolygon(markerType: MapMarkerType) {
        val polygon = mapPolygons.find {
            it.metadata?.getString("polygon") == markerType.name
        }

        polygon?.let { mapPolygon ->
            mapView.mapScene.removeMapPolygon(mapPolygon)
            mapPolygons.removeAll { it.metadata?.getString("polygon") == markerType.name }
        }
    }

    private fun showStartNavigationDialog(
        title: String,
        message: String,
        route: Route,
        isSimulated: Boolean,
    ) {
        val buttonText =
            if (isSimulated) "Start navigation (simulated)"
            else "Start navigation (device location)"
        Logger.log("$title\n$message")

//        val builder: AlertDialog.Builder = Builder(context)
//        builder.setTitle(title)
//            .setMessage(message)
//            .setNeutralButton(buttonText) { dialog, which ->
//                navigationHelper.startNavigation(route, isSimulated, isCameraTrackingEnabled)
//            }
//            .show()
    }

    /**
     * Logs the given title and message, and sends an event with the message to the FCPStreamHandlerPlugin.
     *
     * @param title the title of the error
     * @param message the error message
     */
    private fun onNavigationFailed(title: String, message: String) {
        Logger.log("$title\n$message")

        FCPStreamHandlerPlugin.sendEvent(
            type = FCPChannelTypes.onNavigationFailedFromCarplay.name,
            data = mapOf("message" to message)
        )
//        val builder: AlertDialog.Builder = Builder(context)
//        builder.setTitle(title).setMessage(message).show()
    }

    fun detach() {
        // Disables TBT guidance (if running) and enters tracking mode.
        navigationHelper.stopNavigation()
        // Disables positioning.
        navigationHelper.stopLocating()
        // Disables rendering.
        navigationHelper.stopRendering()
    }
}
